### PR TITLE
[mono] Perform the class init at runtime when static fields of BEFORE_FIELD_INIT type are accessed

### PR DESCRIPTION
**Pull request by LeVladIonescu**
Changed class init checks executed at runtime when accessing static fields of beforefieldinit types, instead of running the cctors during method compilation.

Related to https://github.com/dotnet/runtime/issues/77513.

### CODE REVIEW

**Comment by BrzVlad** [comment]
The commit description is pretty confusing and is referencing code instead of being more of an explanation. Something like: Emit class init checks executed at runtime when accessing static fields of beforefieldinit types, instead of running the cctors during method compilation.

**Comment by EgorBo** [comment]
NOTE: we keep this [path](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Boolean.cs#L196-L217) in Boolean.cs because Mono could trigger ICU initialization because of this beforefieldinit behavior so in theory we can drop that path if you fix this. Although, it might regress perf slightly

**Comment by LeVladIonescu** [comment]
@BrzVlad thanks for pointing it out, updated it.

**Review by vargaz — APPROVED** [review]

**Comment by vargaz** [comment]
This could use a testcase.

**Review by lambdageek — COMMENTED** [review]
does the test project really need to run separately from other tests?

File: src/tests/JIT/Methodical/beforefieldinit/StaticFieldInit.csproj [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <RequiresProcessIsolation>true</RequiresProcessIsolation>
```
[the remark this answers is not in the document]
Comment (LeVladIonescu) [reply src/tests/JIT/Methodical/beforefieldinit/StaticFieldInit.csproj:4]: Yes, you are right. Will modify it.

File: src/tests/JIT/Methodical/beforefieldinit/StaticFieldInit.csproj [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  </PropertyGroup>
+  <PropertyGroup>
+    <DebugType>PdbOnly</DebugType>
+    <Optimize>True</Optimize>
+  </PropertyGroup>
```
[the remark this answers is not in the document]
Comment (LeVladIonescu) [reply src/tests/JIT/Methodical/beforefieldinit/StaticFieldInit.csproj:9]: Just used it from another test which I've added in the past.

**Comment by stephentoub** [comment]
@lambdageek, what should be done with this PR?

**Comment by lambdageek** [comment]
I feel like this is kind of risky to take for .NET 9 at this point.  Static cctors order changes are hard to diagnose downstream.  We could take it up early in .NET 10

**Pull request closed without merging**